File: Java/프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.java
import java.util.HashMap;
import java.util.Arrays;
import java.util.TreeMap;

class Solution {
    private int defaultMoney;
    private int defaultTime;
    private int amountMoney;
    private int amountTime;
    
    public int[] solution(int[] fees, String[] records) {
        defaultTime = fees[0];
        defaultMoney = fees[1];
        amountTime = fees[2];
        amountMoney = fees[3];
        
        HashMap<String, Integer> parkingList = new HashMap<>();
        TreeMap<String, Integer> result = new TreeMap<>();
        
        for (String record : records) {
            String[] recordCategory = record.split(" ");
            String[] timeString = recordCategory[0].split(":");
            int timeInt = timeToInt(Integer.parseInt(timeString[0]), Integer.parseInt(timeString[1]));
            String car = recordCategory[1];
            
            if (recordCategory[2].equals("IN")) {
                parkingList.put(car, timeInt);
            } else {
                int inTime = parkingList.remove(car);
                result.put(car, result.getOrDefault(car, 0) + (timeInt - inTime));
            }
        }
        
        int finalTime = timeToInt(23, 59);
        for (String car : parkingList.keySet()) {
            int inTime = parkingList.get(car);
            result.put(car, result.getOrDefault(car, 0) + (finalTime - inTime));
        }
        
        int[] answer = new int[result.size()];
        int i = 0;
        for (int time: result.values()) {
            answer[i++] = calMoney(time);
        }
        
        return answer;
    }
    
    private int timeToInt(int hour, int min) {
        return (hour * 60) + min;
    }
    
    private int calMoney(int totalTime) {
        if (totalTime <= defaultTime) return defaultMoney;
        
        return defaultMoney + (int) (Math.ceil((double)(totalTime - defaultTime) / amountTime) * amountMoney);
    }
}

File: Swift/프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.swift
import Foundation


var defaultMoney = 0;
var defaultTime = 0;
var amountMoney = 0;
var amountTime = 0;

func solution(_ fees:[Int], _ records:[String]) -> [Int] {
    defaultMoney = fees[1]
    defaultTime = fees[0]
    amountMoney = fees[3]
    amountTime = fees[2]
    
    var parking: [String : Int] = [:]
    var result: [String : Int] = [:]
    
    for record in records {
        let recordCate = record.split(separator: " ").map({ String($0) })
        let timeString = recordCate[0].split(separator: ":").map({ String($0) })
        let timeInt = timeToInt(Int(timeString[0])!, Int(timeString[1])!)
        let car = recordCate[1]
        
        if recordCate[2] == "IN" {
            parking[car] = timeInt
        } else {
            let inTime = parking.removeValue(forKey: car) ?? 0
            result[car] = (result[car] ?? 0) + timeInt - inTime
        }
    }
    
    let finalTime = timeToInt(23, 59)
    for car in parking.keys {
        let inTime = parking[car]!
        result[car] = (result[car] ?? 0) + finalTime - inTime
    }
    
    return result.sorted(by: { Int($0.key)! < Int($1.key)! }).map({ calMoney($0.value) })
}

func timeToInt(_ hour: Int, _ min: Int) -> Int {
    return (hour * 60) + min
}

func calMoney(_ totalTime: Int) -> Int {
    if totalTime <= defaultTime {
        return defaultMoney;
    } else {
        return defaultMoney + Int(ceil((Double(totalTime - defaultTime) / Double(amountTime)))) * amountMoney
    }
}